feat: `codeBlockProvider` improvements

SCWells72 opened the pull request:
Opening this as a draft PR initially to drive discussion, but I'm hoping it will be promoted to a working PR pretty quickly.

This work was driven by some inconsistent behavior and shortcomings I saw with the LSP formatting range-based code block provider, and I figured that by also using the information from the LSP selection range capability, it should be possible to address most if not all of those issues. In practice, that does actually seem to be the case.

This set of changes also includes an update to derive actual language-specific brace pairs from TextMate (`textmate` language) or the IDE's abstract file type (`TEXT` language), but due to a breaking change in the TextMate API in the oldest supported IDE versions, the TextMate aspect is currently commented out. Once support for those oldest versions (the breaking change was made in September, 2023 and it stable in all subsequent IDE versions) has been removed, TextMate support can be restored. Alternatively I could use reflection to support both until there's no longer a need for the older API.

Here's a demo of what these changes enable:

![LSP4IJ_Code_Block_Provider_Demo](https://github.com/user-attachments/assets/606e059d-126e-4d9e-a593-be29048c4e5d)

Specifically the IDE should be able to infer a much more authoritative and accurate set of code blocks based on the current caret position by using selection range information first and folding range information if not found via selection ranges.

The **Move Caret to Code Block Start** and **Move Caret to Code Block End** actions should take you to the correct location based on the current caret. I've enabled the IDE's Presentation Assistant in the demo above so that the OS-specific keybindings for these actions are easily found.

I'll add other details alongside the PR changes themselves.

SCWells72 commented on src/main/java/com/redhat/devtools/lsp4ij/features/codeBlockProvider/LSPCodeBlockProvider.java:
```diff
@@ -0,0 +1,201 @@
+/**
+ * Code block provider that uses information from {@link LSPSelectionRangeSupport} and {@link LSPFoldingRangeBuilder}.
+ */
+public class LSPCodeBlockProvider implements CodeBlockProvider {
```
Now that this is no longer strictly tied to the folding range feature -- which it still uses, but alongside the selection range feature now -- I moved it into its own package. Unfortunately there are enough other changes that Git doesn't see it as a move so there's not a good before/after diff here.

SCWells72 commented on src/main/java/com/redhat/devtools/lsp4ij/features/codeBlockProvider/LSPCodeBlockProvider.java:
```diff
@@ -0,0 +1,201 @@
+
+        // Try to find it first using the selection ranges which tend to be more accurate; we must use the effective
+        // offset for selection ranges to act as if we're in the adjusted braced block
+        TextRange codeBlockRange = getUsingSelectionRanges(
```
Selection ranges from the current offset generally provide more detailed code block information, so try finding the closest containing block using that feature first.

SCWells72 commented on src/main/java/com/redhat/devtools/lsp4ij/features/codeBlockProvider/LSPCodeBlockProvider.java:
```diff
@@ -0,0 +1,201 @@
+        }
+
+        // Failing that, try to find it using the folding ranges
+        return getUsingFoldingRanges(
```
If we couldn't find a good code block from the selection ranges, try to find one using the folding ranges.

SCWells72 commented on src/main/java/com/redhat/devtools/lsp4ij/features/codeBlockProvider/LSPCodeBlockProvider.java:
```diff
@@ -0,0 +1,201 @@
+                    char endChar = documentChars.charAt(endOffset - 1);
+
+                    // If aligned on an open brace and this ends with the expected close brace, use it
+                    if ((startOffset == openBraceOffset)) {
```
The first two conditions here are for when the caret starts exactly on an open or close brace(/bracket/paren) and the evaluated range begins/ends on that offset and ends/begins with the appropriate paired brace. Basically these are an exact brace match.

SCWells72 commented on src/main/java/com/redhat/devtools/lsp4ij/features/codeBlockProvider/LSPCodeBlockProvider.java:
```diff
@@ -0,0 +1,201 @@
+                        }
+                    }
+                    // Otherwise see if it starts and ends with a known brace pair and we'll find the "closest" below
+                    else if ((openBraceOffset == -1) && (closeBraceOffset == -1) && LSPCodeBlockUtils.isCodeBlockStartChar(file, startChar)) {
```
Otherwise we see whether or not the selection range starts and ends with a supported brace pair and, if so, we add it to the list of containing code blocks.

SCWells72 commented on src/main/java/com/redhat/devtools/lsp4ij/features/codeBlockProvider/LSPCodeBlockProvider.java:
```diff
@@ -0,0 +1,201 @@
+            }
+
+            // Return the closest (i.e., smallest) containing text range
+            if (!ContainerUtil.isEmpty(containingTextRanges)) {
```
And then we return the smallest containing code block if at least one was found.

SCWells72 commented on src/main/java/com/redhat/devtools/lsp4ij/features/codeBlockProvider/LSPCodeBlockProvider.java:
```diff
@@ -0,0 +1,201 @@
+                                                   int offset,
+                                                   @Nullable Character openBraceChar,
+                                                   @Nullable Character closeBraceChar) {
+        List<FoldingRange> foldingRanges = LSPFoldingRangeBuilder.getFoldingRanges(file);
```
This is effectively the same logic that was used previously to derive code blocks from folding ranges.

SCWells72 commented on src/main/java/com/redhat/devtools/lsp4ij/features/codeBlockProvider/LSPCodeBlockUtils.java:
```diff
@@ -0,0 +1,141 @@
+ * Utilities for deriving information about code blocks.
+ */
+@ApiStatus.Internal
+public final class LSPCodeBlockUtils {
```
I derived another utility class specifically for code block-related utility methods such as `isCodeBlockStart/EndChar()` and `getCodeBlockStart/EndChar()`. This is also what tries to use the file type-specific information to derive supported brace pairs. It works for both TextMate files and plain text files backed by `AbstractFileType` information, but as stated a few times already, the TextMate aspect is currently commented out due to a breaking API change across the supported IDE versions.

SCWells72 commented on src/main/java/com/redhat/devtools/lsp4ij/features/codeBlockProvider/LSPCodeBlockUtils.java:
```diff
@@ -0,0 +1,141 @@
+            Map<Character, Character> bracePairs = file instanceof TextMateFile ? LSPIJTextMateUtils.getBracePairs(file) :
+                    file.getFileType() instanceof AbstractFileType ? getAbstractFileTypeBracePairs(file) :
+                            null;
+            if (bracePairs == null) {
```
Degrade gracefully to the previous behavior where we assumed that the brace pairs were the default curly braces, square brackets, and parens.

SCWells72 commented on src/main/java/com/redhat/devtools/lsp4ij/features/codeBlockProvider/LSPCodeBlockUtils.java:
```diff
@@ -0,0 +1,141 @@
+
+    @NotNull
+    private static Map<Character, Character> getBracePairsFwd(@NotNull PsiFile file) {
+        return CachedValuesManager.getCachedValue(file, () -> {
```
I've confirmed that the caching here is properly evicted if cached files' file types are changed, e.g., from TextMate to plain text or vice-versa.

SCWells72 commented on src/main/java/com/redhat/devtools/lsp4ij/features/codeBlockProvider/LSPCodeBlockUtils.java:
```diff
@@ -0,0 +1,141 @@
+    }
+
+    @Nullable
+    private static Map<Character, Character> getAbstractFileTypeBracePairs(@NotNull PsiFile file) {
```
This is how supported brace pairs are derived for plain text files backed by `AbstractFileType`s.

SCWells72 commented on src/main/java/com/redhat/devtools/lsp4ij/features/foldingRange/LSPCodeBlockProvider.java:
This was moved to the `codeBlockProvider` package and was already described above.

SCWells72 commented on src/main/java/com/redhat/devtools/lsp4ij/features/foldingRange/LSPFoldingRangeBuilder.java:
```diff
@@ -49,20 +51,11 @@ public class LSPFoldingRangeBuilder extends CustomFoldingBuilder {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(LSPFoldingRangeBuilder.class);
 
-    // NOTE: JetBrains has maintained a long assumption that these are the primary structural block delimiters via
```
Moved into `LSPCodeBlockUtils`.

SCWells72 commented on src/main/java/com/redhat/devtools/lsp4ij/features/foldingRange/LSPFoldingRangeBuilder.java:
```diff
@@ -72,7 +65,7 @@ protected void buildLanguageFoldRegions(@NotNull List<FoldingDescriptor> descrip
         if (!ContainerUtil.isEmpty(foldingRanges)) {
             for (FoldingRange foldingRange : foldingRanges) {
-                TextRange textRange = getTextRange(foldingRange, document);
+                TextRange textRange = getTextRange(foldingRange, file, document);
```
We need to pass the PSI file so that we can derive file type-specific brace pairs.

SCWells72 commented on src/main/java/com/redhat/devtools/lsp4ij/features/foldingRange/LSPFoldingRangeBuilder.java:
```diff
@@ -176,24 +175,6 @@ static TextRange getTextRange(
         return textRange;
     }
 
-    static boolean isOpenBraceChar(@Nullable Character character) {
```
Move into `LSPCodeBlockUtils`.

SCWells72 commented on src/main/java/com/redhat/devtools/lsp4ij/features/foldingRange/LSPFoldingRangeBuilder.java:
```diff
@@ -153,9 +152,9 @@ static TextRange getTextRange(
         if ((startChar != null) && ((openBraceChar == null) || (startChar == openBraceChar))) {
             // If necessary, infer the braces for this block
-            if ((openBraceChar == null) && isOpenBraceChar(startChar)) {
+            if ((openBraceChar == null) && LSPCodeBlockUtils.isCodeBlockStartChar(file, startChar)) {
```
These were moved but also renamed to be less "brace"-specific in their naming.

SCWells72 commented on src/main/java/com/redhat/devtools/lsp4ij/features/selectionRange/LSPExtendWordSelectionHandler.java:
```diff
@@ -71,89 +57,36 @@ public boolean canSelect(@NotNull PsiElement element) {
         }
         return new ArrayList<>(textRanges);
     }
-
```
Moved into `LSPSelectionRangeSupport` so that it can be used by `LSPCodeBlockProvider`.

SCWells72 commented on src/main/java/com/redhat/devtools/lsp4ij/features/selectionRange/LSPSelectionRangeSupport.java:
```diff
@@ -33,12 +51,72 @@
 
+    @NotNull
+    @ApiStatus.Internal
+    public static List<SelectionRange> getSelectionRanges(@NotNull PsiFile file,
```
This was moved from `LSPExtendWordSelectionHandler`.

SCWells72 commented on src/test/java/com/redhat/devtools/lsp4ij/fixtures/LSPCodeBlockProviderFixtureTestCase.java:
```diff
@@ -42,12 +43,18 @@ protected LSPCodeBlockProviderFixtureTestCase(String... fileNamePatterns) {
         }.getType());
         MockLanguageServer.INSTANCE.setFoldingRanges(mockFoldingRanges);
 
+        List<SelectionRange> mockSelectionRanges = JSONUtils.getLsp4jGson().fromJson(mockSelectionRangesJson, new TypeToken<List<SelectionRange>>() {
```
Use the mock selection ranges.

SCWells72 commented on src/main/java/com/redhat/devtools/lsp4ij/features/codeBlockProvider/LSPCodeBlockProvider.java:
```diff
@@ -0,0 +1,251 @@
+                    int startOffset = Math.max(0, textRange.getStartOffset() - 1);
+                    int endOffset = Math.min(documentLength - 1, textRange.getEndOffset());
+                    // These ranges can tend to add whitespace at the end, so trim that before looking for braces
+                    while ((endOffset > startOffset) && Character.isWhitespace(documentChars.charAt(endOffset))) {
```
We sometimes have to perform a pseudo-"trim" on the tail of the folding range to find the closing brace.

SCWells72 commented on src/test/java/com/redhat/devtools/lsp4ij/features/foldingRange/TypeScriptCodeBlockProviderTest.java:
This ends up being a big diff because I had to add mock selection range responses and I added considerably more test cases for more complex code blocks. I'd recommend just taking a look at the test class itself in the IDE instead of here.

SCWells72 commented on src/main/java/com/redhat/devtools/lsp4ij/LSPIJTextMateUtils.java:
```diff
@@ -0,0 +1,92 @@
+ * Utilities for working with TextMate files in LSP4IJ.
+ */
+@ApiStatus.Internal
+public final class LSPIJTextMateUtils {
```
Note that I've hit this exact same issue with 2023.2 in another PR that I'll be submitting shortly. Any chance that you're thinking about dropping support for 2023.2 anytime soon? That would definitely help with some TextMate-related stuff because of these changes between 2023.2 and 2023.3+.

angelozerr replied:
Ok let upgrade that.

Please do that in this PR or an another PR and please update the README also.

SCWells72 replied:
> Ok let upgrade that.
> 
> Please do that in this PR or an another PR and please update the README also.

That upgrade could be a non-trivial amount of work based on whether tests, etc., need to be updated. Let's keep that separate from this work, and I'm happy to do a follow-on PR that enables this logic once LSP4IJ no longer supports 2023.2.

SCWells72 replied:
For what it's worth, I just tried to update the build for 2023.3 instead of 2023.2 in the context of this PR, and while I can build and run the plugin, the tests fail with:

```
Execution failed for task ':instrumentCode'.
> taskdef class com.intellij.ant.InstrumentIdeaExtensions cannot be found
   using the classloader AntClassLoader[]
```

It's possible that the build needs to be updated for the Gradle 2.0 plugin SDK to target higher versions. Either way, it seems like that upgrade should be handled as its own PR.

angelozerr commented on src/main/java/com/redhat/devtools/lsp4ij/LSPIJTextMateUtils.java:
```diff
@@ -0,0 +1,92 @@
+     */
+    @Nullable
+    @ApiStatus.Internal
+    public static Map<Character, Character> getBracePairs(@NotNull PsiFile file) {
```
This method should be removed, right?

SCWells72 replied:
It's up to you. I'd hate to lose it if LSP4IJ is going to be updated to drop support for 2023.2 soon and it should be restored, and it would likely be lost through a squash merge. If you'd like it removed since it's a no-op in its current form, I can do so, but it should hopefully only be commented out for a short time until 2023.2 support is dropped. Let me know what you'd prefer.

angelozerr commented on src/main/java/com/redhat/devtools/lsp4ij/features/codeBlockProvider/LSPCodeBlockProvider.java:
```diff
@@ -0,0 +1,251 @@
+/*******************************************************************************
+ * Copyright (c) 2024 Red Hat, Inc.
```
2025

SCWells72 replied:
This class already existed in 2024 and was moved across packages. It should still be 2024.

angelozerr commented:
@SCWells72 ypu did a lot of changes and it is hard for me to imagine usecases.

Do you think it is possible to write several tests which cover your changes?

SCWells72 commented:
> @SCWells72 ypu did a lot of changes and it is hard for me to imagine usecases.
> 
> Do you think it is possible to write several tests which cover your changes?

I did. Look at the test changes.  They're extensive enough that they don't show up in the diff by default, but you can show them. I roughly doubled the number of test cases with specific examples of the use cases enabled by these changes.

SCWells72 commented:
@angelozerr the PR still shows "Changes requested", but I think I've addressed all feedback. Please let me know if you're waiting for me on something here or have what you need to proceed with the next phase of the review and/or merge.

angelozerr commented:
@SCWells72 could you please record a demo which shows in action the fix that PR provides.

I have tried to set a breakpoint in the LSPCodeBlockProvider and try to format code, try to uses selection ranges, (with Ctrl+W) but I go never in the LSPCodeBlockProvider?

SCWells72 commented:
@angelozerr Apologies for the confusion. This change isn't about formatting. It's about extending the capabilities of the `codeBlockProvider` EP implementation so that it can identify code blocks for the active caret more acccurately and extensively by consulting the selection ranges from the caret first, then the formatting ranges from that same location if a good code block couldn't be identified by selection range.

The best way to understand it is to look at the changes to `TypeScriptCodeBlockProviderTest` where I've added explicit test cases for identification of code blocks that weren't possible with only folding ranges:

https://github.com/redhat-developer/lsp4ij/pull/716/files#diff-5d3b7b59c0f9a335ac1124e7ecb1c29d5a7a46e1e136144c042b931638f5b589

or:

https://github.com/redhat-developer/lsp4ij/blob/8503f91fc919d362e65c9f3537d7f25c85ad601d/src/test/java/com/redhat/devtools/lsp4ij/features/foldingRange/TypeScriptCodeBlockProviderTest.java

If you want to try it within the editor, try with the following:

```typescript
export class Foo {
    bar() {
        let condition: boolean = true;
        if (condition) {
            invokePromise({foo: '', bar: 10})
                .then(() => {
                    console.log('demo');
                })
                .catch((error) => {
                    console.error(error);
                });
        }
    }
}

function invokePromise(params: {foo: string, bar: number;}): Promise<any> {
    console.log(params);
    return Promise.call(params);
}
```

and you can see that the Go To Matching Brace actions work properly for each of the brace pairs including the `.then` and `.catch` callback functions now whereas they didn't with just the formatting ranges-based implementation.

Once this is merged, there are follow-on PRs that add support for improved formatting including client-side on-type formatting using this improved code block information, but this one doesn't have anything to do with formatting.

Let me know if that doesn't clarify the goals/behavior of this PR.

angelozerr commented:
Thanks @SCWells72 !
